### PR TITLE
fix FormationViolation to FormatViolation on CallError (OCPP 2.0.1 Specification, Table B06.FR.17)

### PR DESCRIPTION
**Pull request by dwibudut**
`FormationViolation` to `FormatViolation` base on `OCPP 2.0.1 Specification` Table `B06.FR.17`

### CODE REVIEW

**Comment by dwibudut** [comment]
Hello,

I just fix the `GetInstalledCertificateIdsRequest` and `GetInstalledCertificateIdsResponse` features based on json schema, I've also modified the test function.

Thx

**Comment by dwibudut** [comment]
I added another fix on `typeOfCertificate` to `certificateType` according to json schema in `CertificateSignedRequest' feature

Thx

**Review by lorenzodonini — COMMENTED** [review]
First of all thanks a lot for finding and fixing these implementation errors.

All commits except the first one LG2M (I left one minor comment). 

Regarding the `FormatViolation` I cannot accept the change the way it is, because in v1.6 the error code is indeed called `FormationViolation` (for whatever reason this was changed in v2). I believe this might just be the only change between the two ocppj versions. Fixing one version would break the other. To not break anything we probably need to inject error codes from the top layer. I would suggest splitting this into 2 separate PRs.

File: ocpp2.0.1/v2.go [hunk 1]
```diff
@@ -284,7 +284,7 @@ type CSMS interface {
 	// Retrieves all messages currently configured on a charging station.
 	GetDisplayMessages(clientId string, callback func(*display.GetDisplayMessagesResponse, error), requestId int, props ...func(*display.GetDisplayMessagesRequest)) error
 	// Retrieves all installed certificates on a charging station.
-	GetInstalledCertificateIds(clientId string, callback func(*iso15118.GetInstalledCertificateIdsResponse, error), typeOfCertificate types.CertificateUse, props ...func(*iso15118.GetInstalledCertificateIdsRequest)) error
+	GetInstalledCertificateIds(clientId string, callback func(*iso15118.GetInstalledCertificateIdsResponse, error), certificateTypes []types.CertificateUse, props ...func(*iso15118.GetInstalledCertificateIdsRequest)) error
```
[the remark this answers is not in the document]
Comment (dwibudut) [reply ocpp2.0.1/v2.go:287]: Hello, I've modified these changes based on your review and created a new PR on https://github.com/lorenzodonini/ocpp-go/pull/207

**Comment by dwibudut** [comment]
> First of all thanks a lot for finding and fixing these implementation errors.
> 
> All commits except the first one LG2M (I left one minor comment).
> 
> Regarding the `FormatViolation` I cannot accept the change the way it is, because in v1.6 the error code is indeed called `FormationViolation` (for whatever reason this was changed in v2). I believe this might just be the only change between the two ocppj versions. Fixing one version would break the other. To not break anything we probably need to inject error codes from the top layer. I would suggest splitting this into 2 separate PRs.

For `FormatViolation`, currently I have no idea, whether to just add `FormatViolation` type and keep both (`FormationViolation` and `FormatViolation`) maybe, but the response `CallError` to client will still and always use `FormationViolation` on your top layer `ParseMessage` method

https://github.com/lorenzodonini/ocpp-go/blob/c34f6ad20983d34d7575d9fa05d85368ed0109e2/ocppj/ocppj.go#L373-L390

**Comment by dwibudut** [comment]
New issue https://github.com/lorenzodonini/ocpp-go/issues/210

**Pull request closed without merging**